File: 0042-trapping-rain-water/0042-trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int leftMax = 0;
        int rightMax = 0;

        int left = 0;
        int right = height.length - 1;

        int ans = 0;
        while(left < right){
            if(height[left] < height[right]){
                leftMax = Math.max(leftMax, height[left]);
                ans = ans + leftMax - height[left];
                left++;
            }else{
                rightMax = Math.max(rightMax, height[right]);
                ans = ans + rightMax - height[right];
                right--;
            }
        }
        return ans;
    }
}

File: 0042-trapping-rain-water/0042-trapping-rain-water.py
class Solution(object):
    def trap(self, height):
        """
        :type height: List[int]
        :rtype: int
        """
        left = 0
        right = len(height) - 1
        leftMax = height[left]
        rightMax = height[right]
        total = 0

        while left < right:
            leftMax = max(leftMax, height[left])
            rightMax = max(rightMax, height[right])

            if leftMax < rightMax:
                total += leftMax - height[left]
                left += 1
            else:
                total += rightMax - height[right]
                right -= 1
        
        return total